src/tiered/tiered_handle.c

```diff
@@ -8,63 +8,488 @@
+    dhandle = session->dhandle;
+    tiered = (WT_TIERED *)dhandle;
+    __wt_verbose(session, WT_VERB_TIERED, "TIER_SWITCH: called %s %s", dhandle->name, config);
+
+    /* We might only have a local file tree so far. */
+    WT_RET_NOTFOUND_OK(ret);
+
+    need_local = true;
+    need_object = tiered->tiers[WT_TIERED_INDEX_WRITEABLE] != NULL;
+    need_tree = need_object && tiered->tiers[WT_TIERED_INDEX_READONLY] == NULL;
+    /*
+     * There are four possibilities to our tiers configuration. In all of them we need to create
+     * a new local tier file object dhandle and add it as element index zero of the tiers array.
+     * Then we may or may not do other operations depending on the state otherwise. These are
+     * presented in order of increasing amount of work that needs to be done.
+     *   1. tiers=() - New and empty. We only need to add in the local file object.
+     *   2. tiers=("tier:...") - Existing shared tier only. Here too we only need to add
+     *      in the local file object.
+     *   3. tiers=("file:...", "tier:...") - Both local and shared tiers exist in the metadata.
+     *       We need to create and add the next local file object (N + 1) and create a shared
+     * object in the metadata for the current local file object (N).
+     *   4. tiers=("file:...") - Existing local tier only. We need to do all of the parts listed
+     * in the #3 above, and also create the shared tier metadata entry.
+     *
+     * The order we perform these operations is important. The order is:
+     *   1. Create the "object:" entry in metadata if needed.
+     *   2. Create the "tier:" entry in metadata if needed.
+     *   3. Create the new "file:" local entry in metadata.
+     *   4. Update the "tiered:" with new tiers and object number.
+     *   5. Meta tracking off to "commit" all the metadata operations.
+     *   6. Revise the dhandles in the tiered structure to reflect new state of the world.
+     */
+
+    /*
+     * To be implemented with flush_tier:
+     *    - Close the current object.
+     *    - Copy the current one to the cloud. It also remains in the local store.
+     */
+
+    WT_RET(__wt_meta_track_on(session));
+    tracking = true;
+    /* Create the object: entry in the metadata. */
+    if (need_object)
+        WT_ERR(__tiered_create_object(session, tiered));
+
+    if (need_tree)
+        WT_ERR(__tiered_create_tier_tree(session, tiered));
+
+    if (need_local)
+        WT_ERR(__tiered_create_local(session, tiered));
+
+    /*
+     * Note that removal of overlapping local objects is not in the purview of this function. Some
+     * other mechanism will remove outdated tiers. Here's where it could be done though.
+     */
+
+    /* Update the tiered: metadata to new object number and tiered array. */
+    WT_ERR(__tiered_update_metadata(session, tiered, config));
+    tracking = false;
+    WT_ERR(__wt_meta_track_off(session, true, ret != 0));
+    WT_ERR(__tiered_update_dhandles(session, tiered));
```

Comment:
Does this also need to be wrapped in `WT_SAVE_DHANDLE()`? It looks like `tiered_update_dhandles` might release `session->dhandle`, leaving it `NULL` when it is dereferenced in the following `wt_verbose()` call.